### PR TITLE
ASOAPI: add/remove finalizers

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR adds on to the ASOAPI controllers just enough to manage the appropriate finalizer for each resource.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #4713

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: exp/controllers/azureasomanagedmachinepool_controller.go [hunk 1]
```diff
@@ -172,6 +173,11 @@ func (r *AzureASOManagedMachinePoolReconciler) reconcileNormal(ctx context.Conte
 	defer done()
 	log.V(4).Info("reconciling normally")
 
+	needsPatch := controllerutil.AddFinalizer(asoManagedMachinePool, infrav1exp.AzureASOManagedMachinePoolFinalizer)
```
[the remark this answers is not in the document]
Comment (nojnhuh) [reply exp/controllers/azureasomanagedmachinepool_controller.go:176]: I meant to keep these more or less the same as the existing types but I must have gotten these mixed up. I'll update these.

[the remark this answers is not in the document]
Comment (jackfrancis) [reply exp/controllers/azureasomanagedmachinepool_controller.go:176]: that would explain it!

File: exp/api/v1alpha1/azureasomanagedmachinepool_types.go [hunk 2]
```diff
@@ -20,8 +20,13 @@ import (
...
 )
 
-// AzureASOManagedMachinePoolKind is the kind for AzureASOManagedMachinePool.
-const AzureASOManagedMachinePoolKind = "AzureASOManagedMachinePool"
+const (
+	// AzureASOManagedMachinePoolKind is the kind for AzureASOManagedMachinePool.
+	AzureASOManagedMachinePoolKind = "AzureASOManagedMachinePool"
+
+	// AzureASOManagedMachinePoolFinalizer is the finalizer added to AzureASOManagedMachinePools.
+	AzureASOManagedMachinePoolFinalizer = "azureASOManagedMachinePool.infrastructure.cluster.x-k8s.io"
```
Comment (willie-yao) [inline exp/api/v1alpha1/azureasomanagedmachinepool_types.go]: It looks like other finalizers in CAPZ follow a convention of all lower-case characters. Not sure if this is required but just noting it for consistency: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azurecluster_types.go#L27

File: exp/controllers/azureasomanagedcluster_controller.go [hunk 3]
```diff
@@ -153,5 +172,6 @@ func (r *AzureASOManagedClusterReconciler) reconcileDelete(ctx context.Context,
 	defer done()
 	log.V(4).Info("reconciling delete")
```
Comment (willie-yao) [inline exp/controllers/azureasomanagedcluster_controller.go:173]: nit: Should this be moved to the top of the function, and another log line be added to remove the finalizer? I was just comparing this to the azuremachine controller https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azuremachine_controller.go#L405

File: exp/controllers/azureasomanagedcluster_controller.go [hunk 4]
```diff
@@ -129,6 +133,21 @@ func (r *AzureASOManagedClusterReconciler) reconcileNormal(ctx context.Context,
...
+	}
+	if cluster.Spec.ControlPlaneRef == nil ||
+		cluster.Spec.ControlPlaneRef.APIVersion != infrav1exp.GroupVersion.Identifier() ||
+		cluster.Spec.ControlPlaneRef.Kind != infrav1exp.AzureASOManagedControlPlaneKind {
+		return ctrl.Result{}, reconcile.TerminalError(errInvalidControlPlaneKind)
+	}
+
+	needsPatch := controllerutil.AddFinalizer(asoManagedCluster, clusterv1.ClusterFinalizer)
+	if needsPatch {
+		return ctrl.Result{Requeue: true}, nil
```
Comment (willie-yao) [inline exp/controllers/azureasomanagedcluster_controller.go:148]: I don't see `Requeue: true` being set for any of the other `ctr.Result{}`s. Is this something that's enabled by default or is this case different?

File: exp/api/v1alpha1/azureasomanagedmachinepool_types.go (hunk 2 above)
Comment (nojnhuh) [reply exp/api/v1alpha1/azureasomanagedmachinepool_types.go]: Yep I just fixed this.

File: exp/controllers/azureasomanagedcluster_controller.go (hunk 3 above)
Comment (nojnhuh) [reply exp/controllers/azureasomanagedcluster_controller.go:173]: I don't think that level of consistency exists among all the existing controllers currently. That actually looks like the only place where we log removing the finalizer.

File: exp/controllers/azureasomanagedcluster_controller.go (hunk 4 above)
Comment (nojnhuh) [reply exp/controllers/azureasomanagedcluster_controller.go:148]: In all the other cases we expect some other change to trigger another reconciliation as needed. This change doesn't by default though. Adding the finalizer is basically an intermediate step that we need to persist before finishing a full reconciliation, so this is how we tell the controller to immediately re-reconcile so we do all the stuff that will come after this.

Comment (jackfrancis) [reply exp/controllers/azureasomanagedcluster_controller.go:148]: I was a little confused by this as well. How are the changes to the modified `asoManagedCluster` function parameter object preserved for patching in the next requeue if the immediate requeue signal is sent in the return object?

Comment (nojnhuh) [reply exp/controllers/azureasomanagedcluster_controller.go:148]: We `defer` a function in the top-level `Reconcile` to patch the object after each reconciliation loop which will persist this change, if that's what's not clear.

Comment (nojnhuh) [reply exp/controllers/azureasomanagedcluster_controller.go:148]: The existing controllers do a separate patch inline and keep going within the same reconciliation loop, but this way forces all modification of the object to go through the same patch mechanism at the end, which I feel is slightly more idiomatic.

**Review by willie-yao — APPROVED** [review]
/lgtm

File: exp/controllers/azureasomanagedcluster_controller.go (hunk 4 above)
Comment (jackfrancis) [reply exp/controllers/azureasomanagedcluster_controller.go:148]: defer explains it!

**Comment by jackfrancis** [comment]
squash commits and then lgtm

**Comment by nojnhuh** [comment]
squashed!
/hold cancel

**Pull request merged**